### DISCUSSION TITLE
Physics system update based on frame delta

### DISCUSSION THREAD

**Discussion by MrOnlineCoder** [jrouwe/JoltPhysics#968] [Q&A] [answered]
So far I am using very dumb and straightforward way of performing physics system update, by passing 1 collision step and the frame delta time, and calling it every frame before rendering the scene, without using the multi-threading capabilites of Jolt:
```cpp
m_joltPhysicsSystem.Update(deltaTime, 1, m_tempAllocator, &m_jobThreadPool);
```

And because my FPS is locked to 60 right now, I guess that the reason it works fine. However, I suspect, that when the frame rate will drop, it will affect the simulation respectively.

My question is:
1) Are there any negative effect from the fact that Jolt update is tied to frame delta, in terms of simulation correctness or accuracy? Let's say the frame rate will drop to 30 for few seconds, does it have any impact on the system or performance aside from just the fact of running simulations less frequently?
2) Is there a specfic reason why it's strongly recommended to run simulation at exactly 60 Hz? Are there some wisely picked constants or assumptions in the Jolt code for that, or it would work fine if I run the simulations with a frequency of, let's say, 50 Hz or 30 Hz?

Thanks in advance.

**Comment by jrouwe** [comment, marked as answer]
> Are there any negative effect from the fact that Jolt update is tied to frame delta, in terms of simulation correctness or accuracy? Let's say the frame rate will drop to 30 for few seconds, does it have any impact on the system or performance aside from just the fact of running simulations less frequently?

Yes, the more complicated your scene is, the bigger the difference. In principle you don't want to go below 1/60 th of a second, but you also want the delta time to be as consistent as possible from frame to frame. If you vary delta time, you can clearly see the effect on e.g. a ragdoll as it 'sags' a bit when delta time is big and then becomes 'hard' again when delta time is smaller. Also you may miss collisions completely with bigger delta time as objects can tunnel through other objects.

> Is there a specfic reason why it's strongly recommended to run simulation at exactly 60 Hz? Are there some wisely picked constants or assumptions in the Jolt code for that, or it would work fine if I run the simulations with a frequency of, let's say, 50 Hz or 30 Hz?

So there's not a clear cut off point and it really depends on your scene. In my experience 60 Hz gives a good quality simulation, 50 Hz probably too. 30 Hz in my opinion is too low (but maybe for your scene it will look ok).

**Comment by Avithohol** [comment]
In the same boat. I think what you and me needs is implement interpolation. Bullet physics has btMotionState class, that gives back interpolated transformation for smooth body movements, even on variable framerates. AFAIK Jolt doesn't have such.

See around 'update' how some open source engines implements it:
https://github.com/turanszkij/WickedEngine/blob/1ed5e0d9ac1bf561342a627e78ed101beb4228b3/WickedEngine/wiPhysics_Jolt.cpp#L2699

Quite complex, so I really hope that Jolt provides some helper classes for this case.
